routers/common/middleware.go

```diff
@@ -101,15 +101,20 @@ func stripSlashesMiddleware(next http.Handler) http.Handler {
 }
 
 func Sessioner() func(next http.Handler) http.Handler {
-	return session.Sessioner(session.Options{
-		Provider:       setting.SessionConfig.Provider,
-		ProviderConfig: setting.SessionConfig.ProviderConfig,
-		CookieName:     setting.SessionConfig.CookieName,
-		CookiePath:     setting.SessionConfig.CookiePath,
-		Gclifetime:     setting.SessionConfig.Gclifetime,
-		Maxlifetime:    setting.SessionConfig.Maxlifetime,
-		Secure:         setting.SessionConfig.Secure,
-		SameSite:       setting.SessionConfig.SameSite,
-		Domain:         setting.SessionConfig.Domain,
-	})
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(resp http.ResponseWriter, req *http.Request) {
+			handler := session.Sessioner(session.Options{
+				Provider:       setting.SessionConfig.Provider,
+				ProviderConfig: setting.SessionConfig.ProviderConfig,
+				CookieName:     setting.SessionConfig.CookieName,
+				CookiePath:     setting.SessionConfig.CookiePath,
+				Gclifetime:     setting.SessionConfig.Gclifetime,
+				Maxlifetime:    setting.SessionConfig.Maxlifetime,
+				Secure:         middleware.GetCookieSecure(req),
+				SameSite:       setting.SessionConfig.SameSite,
+				Domain:         setting.SessionConfig.Domain,
+			})
+			handler(next).ServeHTTP(resp, req)
+		})
+	}
```

Comment:
Compiles now. Still not 100% sure this usage is correct, but it appears to work.

Reply:
The "next" call is right, but the Sessioner is not designed to be used like this. eg: It starts a GC goroutine for every call.

The sessioner expects there should be only one instance for all requests.

Reply:
If this is suboptimal, I think the only other option is to let sessioner accept a `function` for the `Secure` attribute as well, or add a `autoSecure` option.

Reply:
It's not only suboptimal, it breaks the Sessioner's design and would cause bugs. The Sessioner also has many internal states, it expects the only one instance do "Init" (eg: connect to redis) and do session management (eg: Lock, GC).

Although we could make some changes to make a Sessioner work with dynamic Cookie Secure option, but I think it's too complex to do that, and "dynamic Cookie Secure" doesn't seem really bring enough benefit.

The mentioned "AppURL + COOKIE_SECURE option" should be simple enough and I don't see any problem by doing that.

Reply:
It's not just cookie secure. If the backend wants for example to reconstruct the original URL a client used (for example for audit log), XFP is required for that as well.

Maybe I'll look into sessioner accepting a function (typecheck with Reflect), that seems the simplest option.

Reply:
Unless the app runs on multiple URLs :)

Reply:
> Unless the app runs on multiple URLs :)

Yup, but the AppURL should also be correct in this case, otherwise the nofication/mail/webhook would stop working.

As long as the AppURL should be always right, then just use it.

Reply:
Opened https://gitea.com/go-chi/session/issues/7.